### ISSUE TITLE
Generated types for enums fail type checks

### ISSUE THREAD

**Issue by robbrit** [dropbox/mypy-protobuf#219]
We're using protobuf enums all over the place, and with a recent upgrade of mypy-protobuf we're suddenly seeing all sorts of errors like this:

```
error: Incompatible types in assignment (expression has type "V", variable has type "JobType")
```
The code is simply using the protobufs:
```python
class JobStatus(NamedTuple):
    # ...
    job_type: jobs_pb2.JobType = jobs_pb2.JOB_UNKNOWN
```
In this particular case it's setting a default value on a `NamedTuple`, but we get the same error for any place in the code that uses a protobuf enum: function calls, variable assignments, etc.

Is there an alternative approach, rather than using `# type: ignore` or rolling back to an older version of `mypy-protobuf`?

**Comment by nipunn1313** [comment]
Hi. I’m afk for a few days, but wanted to post quickly from my phone.

This looks like one of the key non backward compat changes to how enums are generated in 2.0 - which should be documented in both the release notes and on the main readme.

tl;dr is that JobStatus as generated is an EnumTypeWrapper and not an int. You likely want to type it as an int or as JobStatus.V

the recommended approach is to type the enum values correctly (int or JobStatus.V), rather than using the enumtypewrapper (which incorrectly used to work in 1.*)

if you have documentation or self discovery suggestions that would make this more self explanatory - Id be excited to hear. This is absolutely a major point of confusion

**Comment by nipunn1313** [comment]
See https://github.com/dropbox/mypy-protobuf#types-enum-int-values-more-strongly

and 
https://github.com/dropbox/mypy-protobuf/blob/master/CHANGELOG.md#121

Appears to have landed in 1.21 - apologies, definitely should have landed it in major release 2.0 for its non compatibility, but just going to take that lesson going forward to prioritize stability.

**Comment by robbrit** [comment]
So to ensure I understand correctly: the protobuf compiler is producing them as integers, which reduces type safety because you'd be able to assign any enum to any enum?

And to use this correctly we would replace all instances of `JobType` in type annotations with `"JobType.V"` (quotes required).

**Comment by nipunn1313** [comment]
I believe your understanding is correct.

One small nit - you’ll want to edit any instance of JobType being used to
type an int Value.

You wouldn’t want to change JobType being used correctly as an
EnumTypeWrapper (eg JobType.Name(foo) or JobType.FOO

In dropbox’s codebase, nearly all instances were of the first variety.
If you have at least Python 3.9 or use “from __future__ import
annotations”, you may be able to avoid needing the quotation marks.
—Nipunn

**Comment by robbrit** [comment]
Diving back into this one, we ended up de-prioritizing the upgrade due to this until now.

I don't think I understand correctly anymore. In no cases are we using `int` to specify an enum type, we're always using the name of the enum: `JobType`. They are typically either fields on a `NamedTuple` or parameters to a function call.

We'll have things like this:
```python
def get_jobs(job_type: jobs_pb2.JobType) -> Iterable[Job]:
   ...
```
The problem is that anything that calls `get_jobs()` fails, even if we pass values that are of the `JobType` enum.

It fixes it if we change the declaration to this:
```python
def get_jobs(job_type: "jobs_pb2.JobType.V") -> Iterable[Job]:
```
Is there a way to get around this? Having to rewrite function signatures and class definitions across our entire codebase would be a bit problematic.

**Comment by nipunn1313** [comment]
Hi @robbrit - thanks for the detailed explanation and interest.

If you look at code generated by the protobuf python compiler (`{file}_pb2.py`) - you'll find that `jobs_pb2.JobType` is an `EnumTypeWrapper`, which is NOT an instance of the enum. See the [docs](https://developers.google.com/protocol-buffers/docs/reference/python-generated#enum). Instances of the enum are `int`s, which mypy-protobuf types as `jobs_pb2.JobType.V`.

I personally find this to have been a confusing choice by the google protobuf developers - especially because regular python enums have the name of the enum as instances of the enum. However, it is what it is.

You are likely running into a migration issue from the [backward incompatible changes](https://github.com/dropbox/mypy-protobuf/blob/main/CHANGELOG.md#20) from mypy-protobuf 1.* -> 2.*.
In 1.*, mypy-protobuf used to generate `JobType` incorrectly as BOTH an `EnumTypeWrapper` AND an `int`. With 2.*, they are typed correctly as `JobType` and `int` (represented as `JobType.V`).

I empathize with the challenge you present. I corrected all the function signatures across the DBX codebase and there were thousands. If you are able to just slam it out in one go, that's the easiest way. I was not able to do that, but I was able to migrate `.proto` files in groups by straddling our `protoc` invocations across multiple versions of mypy-protobuf - and over the course of a few days and 10-20 diffs, I was able to bang them all out.

The general strategy I used was to move a handful of `.proto` files over to the newer mypy-protobuf. I'd then look for all instances of affected enums (eg `JobType`) and determine whether `JobType` was being used as an enum value `JobType.V` or as an EnumTypeWrapper (eg `JobType.keys()`). A few sed substitutions and fixups went a long way.

One note, is that if you use `__from__ future import annotations`, then you won't need quotation marks. There are docs on this here - https://github.com/dropbox/mypy-protobuf#types-enum-int-values-more-strongly

Hopefully this is helpful!
